### PR TITLE
Add LivenessThreshold

### PR DESCRIPTION
**Pull request by hmoog**

### CODE REVIEW

**Review by piotrm50 — APPROVED** [review]
Let's merge this!

File: pkg/core/promise/promise.go [hunk 1]
```diff
@@ -5,20 +5,21 @@ import (
 
 	"github.com/iotaledger/hive.go/ds/orderedmap"
 	"github.com/iotaledger/hive.go/runtime/syncutils"
+	"github.com/iotaledger/iota-core/pkg/core/types"
```
[the remark this answers is not in the document]
Comment (hmoog) [reply pkg/core/promise/promise.go:8]: Yeah, but let's do it in another PR as the hive.go dependencies are finally merged :p

File: pkg/protocol/engine/mempool/v1/state_metadata.go [hunk 2]
```diff
@@ -69,7 +69,7 @@ func (s *StateMetadata) State() mempool.State {
 	return s.state
 }
 
-func (s *StateMetadata) ConflictIDs() *lpromise.Set[iotago.TransactionID] {
+func (s *StateMetadata) ConflictIDs() reactive.Set[iotago.TransactionID] {
```
Comment (jkrvivian) [inline pkg/protocol/engine/mempool/v1/state_metadata.go:72]: `reactive.DerivedSet`?

Comment (muXxer) [reply pkg/protocol/engine/mempool/v1/state_metadata.go:72]:
```suggestion
func (s *StateMetadata) ConflictIDs() reactive.DerivedSet[iotago.TransactionID] {
```
I agree, that's confusing. It should be the derived set, right?

File: pkg/protocol/engine/mempool/v1/transaction_metadata.go [hunk 3]
```diff
@@ -97,31 +97,31 @@ func (t *TransactionMetadata) Transaction() mempool.Transaction {
...
+	outputs := ds.NewSet[mempool.StateMetadata]()
 	for _, output := range t.outputs {
 		outputs.Add(output)
 	}
 
 	return outputs
 }
 
-func (t *TransactionMetadata) ConflictIDs() *lpromise.Set[iotago.TransactionID] {
+func (t *TransactionMetadata) ConflictIDs() reactive.Set[iotago.TransactionID] {
```
Comment (muXxer) [inline pkg/protocol/engine/mempool/v1/transaction_metadata.go:124]:
```suggestion
func (t *TransactionMetadata) ConflictIDs() reactive.DerivedSet[iotago.TransactionID] {
```

File: pkg/protocol/engine/tipmanager/v1/tip_metadata.go [hunk 4]
```diff
@@ -15,341 +13,211 @@ type TipMetadata struct {
...
-	isWeakTip *lpromise.Value[bool]
+	isWeakTip reactive.Variable[bool]
+
+	// isMarkedOrphaned is true if the liveness threshold has been reached and the block was not accepted.
+	isMarkedOrphaned reactive.Variable[bool]
+
+	// isOrphaned is true if the block is either strongly or weakly orphaned.
+	isOrphaned reactive.Variable[bool]
+
+	// anyStrongParentStronglyOrphaned is true if the block has at least one orphaned parent.
```
Comment (muXxer) [inline pkg/protocol/engine/tipmanager/v1/tip_metadata.go:61]:
```suggestion
	// anyStrongParentStronglyOrphaned is true if the block has at least one strong parent that is strongly orphaned.
```

File: pkg/protocol/engine/tipmanager/v1/tip_metadata.go [hunk 5]
```diff
@@ -15,341 +13,211 @@ type TipMetadata struct {
...
-	})
-
-	t.orphanedStrongParents.OnUpdate(func(_, orphanedStrongParents int) {
-		t.isOrphaned.Compute(func(_ bool) bool {
-			return orphanedStrongParents > 0 || t.isMarkedOrphaned.Get()
-		})
-	})
+	// unsubscribe when the parent is evicted, since we otherwise continue to hold a reference to it.
+	unsubscribe := strongParent.stronglyConnectedStrongChildren.Monitor(t.isStronglyConnectedToTips)
+	strongParent.evicted.OnUpdate(func(_, _ bool) { unsubscribe() })
```
Comment (muXxer) [inline pkg/protocol/engine/tipmanager/v1/tip_metadata.go:210]: would this also trigger the unsubscribe if the strongParent is already evicted in the moment we call `OnUpdate` ?
Is this a possible memory leak?

File: pkg/protocol/engine/tipmanager/v1/tip_metadata.go [hunk 6]
```diff
@@ -15,341 +13,211 @@ type TipMetadata struct {
...
-// setupWeakParent sets up the parent and children related properties for a weak parent.
-func (t *TipMetadata) setupWeakParent(weakParent *TipMetadata) {
-	weakParent.OnEvicted( // unsubscribe on eviction of the parent (prevent memory leaks).
-		t.isConnectedToTips.OnUpdate(func(_, isConnectedToTips bool) {
-			weakParent.connectedWeakChildren.Compute(lo.Cond(isConnectedToTips, increase, decrease))
-		}),
-	)
+	// unsubscribe when the parent is evicted, since we otherwise continue to hold a reference to it.
+	unsubscribe := weakParent.connectedWeakChildren.Monitor(t.isConnectedToTips)
+	weakParent.evicted.OnUpdate(func(_, _ bool) { unsubscribe() })
```
Comment (muXxer) [inline pkg/protocol/engine/tipmanager/v1/tip_metadata.go:219]: same as above.

File: pkg/protocol/engine/tipmanager/v1/tip_metadata.go (hunk 5 above)
Comment (piotrm50) [reply pkg/protocol/engine/tipmanager/v1/tip_metadata.go:210]: Yes. If `evicted` has been triggered before this line, then the callback is executed immediately, so unsubscribe will be called in the described case.

Comment (muXxer) [reply pkg/protocol/engine/tipmanager/v1/tip_metadata.go:210]: ok cool, then this is resolved! :+1:

File: pkg/protocol/engine/mempool/v1/state_metadata.go (hunk 2 above)
Comment (piotrm50) [reply pkg/protocol/engine/mempool/v1/state_metadata.go]: Changed to ReadableSet, so that callers can't modify the contents of the set.

**Review by muXxer — APPROVED** [review]

File: pkg/protocol/engine/mempool/v1/state_metadata.go (hunk 2 above)
Comment (muXxer) [reply pkg/protocol/engine/mempool/v1/state_metadata.go:72]: Ok lets change that in another PR where we cleanup all sets to use ReadableSets where necessary.

**Pull request merged**